torch/fx/passes/split_module.py

```diff
@@ -207,14 +279,22 @@ def instantiate_node_partition_mapping(node):
             )
             continue
 
-        instantiate_node_partition_mapping(node)
+        if assert_monotonically_increasing:
+            pid = split_callback(node)
+            assert highest_partition <= pid,\
+                ("autocast or set_grad_enabled require monotonically increasing partitions:"
+                 f"highest: {highest_partition}, this node's: {pid}")
+            highest_partition = pid
 
-        torch.fx.graph.map_arg(
-            node.args, lambda def_node: record_cross_partition_use(def_node, node)
-        )
-        torch.fx.graph.map_arg(
-            node.kwargs, lambda def_node: record_cross_partition_use(def_node, node)
-        )  # noqa: B950
+        # do not capture cross-partition dependencies for global state nodes as they will be
+        # self-contained - their setup and unwind will be isolated to each partition submodule.
+        if node.target not in GLOBAL_STATE_NODES:
```

Comment:
i'm not clear on the reasons for needing this.

1- is there a 'use' relationship between set_grad_enabled node and some other fx nodes that come after it in the graph?
2- when we duplicate `set_grad_enabled` node at the beginning of other subgraphs, wouldn't we update the graph so that nodes in partition B only depend on partition B's `set_grad_enabled`?

given that supposition, i might claim we can safely record_cross_partition_use for all nodes.  (perhaps, we need to move that 'prelude' section up above record_cross_partition_use first though?

Reply:
The particular thing this handles is that the way that the autocast._exit is implemented as passing the result of autocast._enter to it. This introduces a cross-partition dependency. However, downstream compilers neither know how to deal with passing the autocast object as an argument, nor handling the case where autocast state is mutated by the graph. 

That being said, I think we don't actually need to forbid cross-partition the arguments to autocast._enter and set_grad_enabled - as much as they are likely to be constants anyway. 

There could be a subtle correctness issue if we don't handle the case where they are not constants I.e. data dependent - and the value originates in some other subgraph. I'm not sure if this is possible to create a test case for as dynamo might enforce that these arguments are coerced into constants by guarding, but I could try to force the value to be a symint.

Reply:
One such way would be to create a boolean tensor with 1 elem and call .item on it

Reply:
Actually, let's not add a test. Looking a the dynamo code, there is no possibility of the args being non-constant.

Reply:
I've updated to only exclude autocast._exit

Reply:
i think i still dont get why exit depends on enter from another graph.  if you are in case 2 (need to add both enter and exit to a middle graph) then your new exit would depend only on your new enter, right?

and in case 3, the tail graph, where you add a new enter and there was an existing exit - wouldn't you update the old exit to reference the new enter, or, create a new exit too?

my point was that there shouldn't be any cross-graph deps related to autocast/amp after your work is done, so, the cross-graph validation should just be allowed to run normally.

Reply:
Right, the problem is that the per-subgraph setup/unwind is not instantiated until later. We are still at the preprocessing stage. If we did include autocast._exit in the dependency recording, then we wrongly identify a dependency between the partition where it is setup and the partition where it is unwound.

Reply:
yea thats what i meant by this comment

> (perhaps, we need to move that 'prelude' section up above record_cross_partition_use first though?

e.g. maybe we should first install the new enters/exits, then do teh dependency analysis.

However i'm not sure it matters.  Maybe its just fine either way.

Reply:
> e.g. maybe we should first install the new enters/exits, then do teh dependency analysis.

Yeah, I don't think we want that, as the record cross-partition use is a preprocessing stage, similar to how the global state nodes paritition dependency info gathering is a preprocessing stage, where we collect information about how we want to transform the graph in the graph transformation stage.

Reply:
I'm going to call this resolved, basically, we do this as we take a different path for preprocessing the dependencies for autocast._exit